### PR TITLE
Refactor ActOnArgs.kron/factor/transpose to reduce code duplication

### PR DESCRIPTION
**Pull request by daxfohl**
Step 3 for feedforward, outlined in https://tinyurl.com/cirq-feedforward.

This PR reduces duplication by moving redundant code from ActOnArgs.kron etc, such as `qubits = self.qubits + other.qubits` into the ActOnArgs base class, and then using handlers in subclasses to append subclass-specific information. It also updates these methods with an `inplace` option to allow them to be consistent with the swap/rename methods added in #4169

@95-martin-orion has the most context here.

### CODE REVIEW

**Review by MichaelBroughton — COMMENTED** [review]
This looks good. Just a few high level questions in some areas.

File: cirq-core/cirq/sim/act_on_args.py [hunk 1]
```diff
@@ -105,31 +108,72 @@ def _perform_measurement(self, qubits: Sequence['cirq.Qid']) -> List[int]:
...
 
-    @abc.abstractmethod
     def copy(self: TSelf) -> TSelf:
         """Creates a copy of the object."""
+        args = copy.copy(self)
+        self._on_copy(args)
+        args._log_of_measurement_results = self.log_of_measurement_results.copy()
+        return args
+
+    def _on_copy(self: TSelf, args: TSelf):
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/sim/act_on_args.py:118]: I was inclined at first too, but there are a couple reasons not to. Backwards compatibility: right now anyone who has implemented `copy` manually still works; if we make this abstract then it'll fail to instantiate. Second it's unnecessary overhead on unit tests that don't use it. Third it's also unnecessary overhead for some of the functions like factor/kron that are optional.

File: cirq-core/cirq/sim/act_on_args.py [hunk 2]
```diff
@@ -105,31 +108,72 @@ def _perform_measurement(self, qubits: Sequence['cirq.Qid']) -> List[int]:
...
 
     def create_merged_state(self: TSelf) -> TSelf:
         """Creates a final merged state."""
         return self
 
-    def kronecker_product(self: TSelf, other: TSelf) -> TSelf:
+    def kronecker_product(self: TSelf, other: TSelf, *, inplace=False) -> TSelf:
         """Joins two state spaces together."""
-        raise NotImplementedError()
+        args = self if inplace else copy.copy(self)
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/sim/act_on_args.py:128]: It's shallow by design; the thing it's replacing was also a shallow operation. Deep would copy the whole state vector and be a perf regression. If there's anything in the subclass that needs copied over, then that can be added to the corresponding _on method or __copy__ can be overridden.

**Review by MichaelBroughton — APPROVED** [review]
LGTM

**Pull request merged**